### PR TITLE
Docs: Add Grafana Integration Guide

### PR DESCRIPTION
**Pull request by alexfornuto**
## Summary

Adds a guide for adding a Grafana route with automatic login using Pomerium's JWT.

## Related issues

Closes https://github.com/pomerium/docs/issues/56

Blocked to merge by https://github.com/pomerium/pomerium/pull/2667 - This guide references and links to content in this PR.

## Checklist

- [ ] reference any related issues
- [x] updated docs
- [ ] updated unit tests
- [ ] updated UPGRADING.md
- [x] add appropriate tag (`improvement` / `bug` / etc)
- [x] ready for review

### CODE REVIEW

File: docs/guides/grafana.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    ingress.pomerium.io/policy: '[{"allow":{"and":[{"domain":{"is":"example.com"}}]}}]'
+    #ingress.pomerium.io/secure_upstream: true # Uncomment if Grafana is serving content over HTTPS
+    ingress.pomerium.io/pass_identity_headers: true
+spec:
+  ingressClassName: pomerium
+  rules:
+  - host: grafana.localhost.pomerium.io
+    http:
+      paths:
+      - backend:
```
Comment (travisgroth) [inline docs/guides/grafana.md:76]:
```suggestion
      - path: /
        pathType: Prefix
        backend:
```

**Review by travisgroth — CHANGES_REQUESTED** [review]
Left some suggestions on the non-kubernetes path. 

For kubernetes there's some changes needed to the Ingress resource.

File: docs/guides/grafana.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+- defines the header to look at to provide the JWT,
+- associates the email_claim in the JWT with the email of the Grafana user
+- specifies the location of the signing key for the JWT to validate it,
+- sets a 60 minute cache time for the token.
+
+Once you've saved an exited the file, restart Grafana.
+
+## Add Users to Grafana
+
+At this stage Pomerium is configured to route traffic to Grafana and include identity information via the JWT, and Grafana is configured to use the `email` claim in the JWT to associate an incoming connection with a user. But the user must still exist in Grafana to be associated. Otherwise, you will see this error in the browser after authenticating:
```
[the remark this answers is not in the document]
Comment (alexfornuto) [reply docs/guides/grafana.md]: Good point, addressed in 1be4f0c78bd955ae35ee7277acc6174814c699b1

File: docs/guides/grafana.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+```json
+{"message":"Invalid JWT"}
+```
+
+1. To add users without requiring them to accept an invitation, log in to Grafana directly as an admin user.
```
[the remark this answers is not in the document]
Comment (alexfornuto) [reply docs/guides/grafana.md]: Done in 1be4f0c78bd955ae35ee7277acc6174814c699b1

File: docs/guides/grafana.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+In instances where Grafana cannot get the signing key for the JWTs from the Pomerium authenticate service, you can place a copy of the key locally.
+
+For example, wildcard certificates signed by LetsEncrypt may still be cross-signed by the [expired DST R3 root]. While many browsers still trust these certificates (as long as they are also signed by a valid root), some applications reject them, including Grafana:
+
+```log
+logger=context error=Get "https://authenticate.localhost.pomerium.io/.well-known/pomerium/jwks.json": x509: certificate signed by unknown authority
+```
+
+To circumvent this issue, you can use `curl` or `wget` to download the signing key locally, and the `jkw_set_file` key to provide it to Grafana:
```
[the remark this answers is not in the document]
Comment (alexfornuto) [reply docs/guides/grafana.md]: Excellent idea, and offers room for better instructions. Applied in f39fe2d42ba46e7096936b4d33ed1cd84dff9432

File: docs/guides/grafana.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+```bash
+curl https://authenticate.localhost.pomerium.io/.well-known/pomerium/jwks.json > /etc/grafana/jwks.json
+```
+::::
+:::: tab wget
+From the Grafana host:
+```bash
+cd /etc/grafana/
+wget https://authenticate.localhost.pomerium.io/.well-known/pomerium/jwks.json
```
Comment (travisgroth) [inline docs/guides/grafana.md:179]:
```suggestion
wget -O /etc/grafana/jwks.json https://authenticate.localhost.pomerium.io/.well-known/pomerium/jwks.json
```

**Review by travisgroth — APPROVED** [review]
Two suggestions but LGTM

File: docs/guides/grafana.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+[Grafana](https://grafana.com/) is an open-source analytics and monitoring solution. The many user-contributed [Dashboards](https://grafana.com/grafana/dashboards/) make it popular for enthusiasts as well as professionals.
+
+This guide will demonstrate how to secure an instance of Grafana behind Pomerium, and provide users with a seamless login to Grafana using your Identity Provider (**IdP**).
+
+## Before You Begin
+
+This guide begins with the following steps assumed complete:
+
+- A running instance of Pomerium. If you haven't already installed Pomerium, see our articles for installation on [Docker](/docs/install/readme.md), [Kubernetes](/docs/install/helm.md), or as an all-in-one [binary](/docs/install/binary.html).
+- Administrator access to a working Grafana instance, including write access to the config file `grafani.ini`, usually found in `/etc/grafana`.
```
Comment (cmo-pomerium) [inline docs/guides/grafana.md:24]: check spelling? is it grafini.ini?

**Review by cmo-pomerium — COMMENTED** [review]
Left comments highlighting things I weren't sure of and made copy editing changes

File: docs/guides/grafana.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+This guide uses the following temporary values in steps and configuration examples, which will need to be adjusted for your setup:
+
+- `http://grafana:3000` - The path Pomerium will use to access Grafana. This example emulates a common Docker-based setup.
+- `http://grafana.local:3000` - The path to access the Grafana interface from your local computer. We will need direct access to add at least one user before Pomerium is configured.
+- `https://grafana.localhost.pomerium.io` - The path to access Grafana through Pomerium. Change this to match the domain space Pomerium is configured in.
+
+## Enable JWT Authentication in Grafana
+
+Edit `grafana.ini` to configure [JWT authentication]. Replace `auth.localhost.pomerium.io` with the value of [`authenticate_service_url`] in your Pomerium configuration:
```
Comment (cmo-pomerium) [inline docs/guides/grafana.md:34]: Do we want to define JWT acronym?

**Review by cmo-pomerium — COMMENTED** [review]
Added comment on acronym

File: docs/guides/grafana.md (hunk 6 above)
Comment (alexfornuto) [reply docs/guides/grafana.md:24]: @cmo-pomerium where are you seeing `grafini.ini`? Or are you saying maybe it _should_ be `grafini.ini`? If so, I can confirm it's definitely `grafani.ini`.

File: docs/guides/grafana.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+This guide will demonstrate how to secure an instance of Grafana behind Pomerium, and provide users with a seamless login to Grafana using your Identity Provider (**IdP**).
+
+## Before You Begin
+
+This guide begins with the following steps assumed complete:
+
+- A running instance of Pomerium. If you haven't already installed Pomerium, see our articles for installation on [Docker](/docs/install/readme.md), [Kubernetes](/docs/install/helm.md), or as an all-in-one [binary](/docs/install/binary.html).
+- Administrator access to a working Grafana instance, including write access to the config file `grafani.ini`, usually found in `/etc/grafana`.
+
+This guide uses the following temporary values in steps and configuration examples, which will need to be adjusted for your setup:
```
[the remark this answers is not in the document]
Comment (alexfornuto) [reply docs/guides/grafana.md:26]: Your initial thought was correct. The language here was set up to specify that it's not just the config examples, but the commands and copy that use these placeholder variables. But maybe I'm being overly cautious with that thinking. How about:

```suggestion
This guide uses the following temporary values in commands and configuration examples , which will need to be adjusted for your setup:
```

File: docs/guides/grafana.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+## Enable JWT Authentication in Grafana
+
+Edit `grafana.ini` to configure [JWT authentication]. Replace `auth.localhost.pomerium.io` with the value of [`authenticate_service_url`] in your Pomerium configuration:
+
+```ini
+[auth.jwt]
+enabled = true
+header_name = X-Pomerium-Jwt-Assertion
+email_claim = email
+jwk_set_url = https://auth.localhost.pomerium.io/.well-known/pomerium/jwks.json
```
[the remark this answers is not in the document]
Comment (alexfornuto) [reply docs/guides/grafana.md:41]: Yes, for Java Web Key. See https://grafana.com/docs/grafana/latest/auth/jwt/#verify-token-using-a-json-web-key-set-loaded-from-https-endpoint

File: docs/guides/grafana.md (hunk 7 above)
Comment (alexfornuto) [reply docs/guides/grafana.md:34]: We do a bit lower, on [line 47](https://github.com/pomerium/pomerium/pull/2742/files#diff-c9db3551ad9e3b330e2cc09dd55e72febbeef768ae4048045ee596c2e73c4fc3R47)

**Comment by cmo-pomerium** [comment]
Followed up on the review, looks good to me!

File: docs/guides/grafana.md (hunk 6 above)
Comment (alexfornuto) [reply docs/guides/grafana.md:24]: OK, I don't know what I was thinking last night. We were both wrong.

```suggestion
- Administrator access to a working Grafana instance, including write access to the config file `grafana.ini`, usually found in `/etc/grafana`.
```

File: docs/guides/grafana.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    content: >-
+      pomerium identity-access-proxy data logging graphing grafana
+      authentication authorization
+description: >-
+  This guide covers how to use Pomerium to authenticate and authorize users of Grafana.
+---
+
+# Securing Grafana with Pomerium
+
+[Grafana](https://grafana.com/) is an open-source analytics and interactive visualization web application monitoring solution. It provides many user-contributed [Dashboards](https://grafana.com/grafana/dashboards/) that make it popular for enthusiasts as well as professionals.
```
Comment (alexfornuto) [inline docs/guides/grafana.md:15]: > [Grafana](https://grafana.com/) is an open-source analytics and interactive visualization web application monitoring solution.

@cmo-pomerium This got a little long-winded, eh? How about:

```suggestion
[Grafana](https://grafana.com/) is an open-source and interactive analytics visualization  and monitoring web application. It provides many user-contributed [Dashboards](https://grafana.com/grafana/dashboards/) that make it popular for enthusiasts as well as professionals.
```

**Review by cmo-pomerium — COMMENTED** [review]
Looks good to me.

**Pull request merged**